File: backtracking/PalindromePartitioning.cpp
bool isPalindrome(string A) {
    if (A.size() <= 1) return true;
    return (A[0] == A[A.size() - 1] && isPalindrome(A.substr(1, A.size() - 2)));
}

vector<vector<string>> Solution::partition(string A) {
    vector<vector<string>> output;
    if (A.size() <= 1) {
        vector<string> temp;
        if (A.size() == 1) temp.push_back(A);
        output.push_back(temp);
        return output;
    }
    int len = 1;
    while (len <= A.size()) {
        string cur = A.substr(0, len);
        if (isPalindrome(cur)) {
            vector<vector<string>> temp = partition(A.substr(len, A.size() - len));
            for (int i = 0; i < temp.size(); i++) {
                temp[i].insert(temp[i].begin(), cur);
            }
            output.insert(output.end(), temp.begin(), temp.end());
        }
        len++;
    }
    return output;
}

File: AlgoMonster/backtracking/palindrome_partitioning.py
from typing import List

def is_palindrome(s: str) -> bool:
    if len(s) <= 1:
        return True
    return s[0]==s[-1] and is_palindrome(s[1:-1]) 

def partition(s: str) -> List[List[str]]:
    def dfs(path, cur, list):
        if len(cur) <= 1:
            if len(cur) == 1:
                path.append(cur)
            list.append(path[:])
            if len(cur) == 1:
                path.pop()
            return
        for i in range(1, len(cur) + 1):
            substr = cur[:i]
            if is_palindrome(substr):
                path.append(substr)
                dfs(path, cur[i:], list)
                path.pop()
    list = []
    dfs([], s, list)
    return list

if __name__ == '__main__':
    s = input()
    res = partition(s)
    for row in res:
        print(' '.join(row))
